src/symbolic.jl

```diff
@@ -142,171 +130,104 @@ end
     end
 
+    factor, powers
 end
 
-isvar(x::Symbolic, vars::Vector) = any(var -> isequal(x, var), vars)
+function split_factor(expr::Div, iv, smap, vars)
+    num, denom = arguments(expr)
+    isconstant(denom, vars, iv) || error("The denominator $denom is not constant.")
+    
+    factor, powers = split_factor(num, iv, smap, vars)
+    
+    factor / denom, powers
+end
 
-function extract_pwr!(powers::Vector, expr::Symbolic, smap::Dict, vars::Vector)
-    args = arguments(expr)
-    if isvar(args[1], vars) && args[2] isa Int && args[2] >= 0
-        idx = smap[args[1]]
-        if powers[idx] != 0
-            error("$args[1] occurring multiple times in a monomial is unexpected. In $expr")
-        else
-            powers[idx] = args[2]
-        end
+function split_factor(expr, iv, smap, vars) 
+    if isconstant(expr, vars, iv)
+        expr, zeros(length(vars))
+    elseif isvar(expr, vars)
+        1, map(isequal(expr), vars)
     else
-        error("Unexpected term: $expr")
+        error("Expression $expr could not be parsed correctly!")
     end
 end
 
-function extract_mul(expr::Symbolic, smap::Dict, vars::Vector, iv::Sym)
+function polynomial_propensity(expr::Div, iv, smap, vars)
+    num, denom = arguments(expr)
+    isconstant(denom, vars, iv) || error("The denominator $denom is not constant.")
+    factors, powers = polynomial_propensity(num, iv, smap, vars)
+    factors ./ denom, powers
+end
 
-    powers = zeros(Int, length(smap))
+function polynomial_propensity(expr::Add, iv, smap, vars)
     factors = []
-
-    for arg in arguments(expr)
-        if isconstant(arg, vars, iv)
-            push!(factors, arg)
-        elseif operation(arg) == ^
-            extract_pwr!(powers, arg, smap, vars)
-        elseif isvar(arg, vars)
-            powers[smap[arg]] = 1
-        else
-            error("Unexpected operation: $(operation(arg)) in $expr")
+    powers = Vector{Int}[]
+    
+    for term in arguments(expr)
+        factor_term, power_term = try
+            split_factor(term, iv, smap, vars)
+        catch e
+            error("Propensity function $term is non-polynomial? \n" * string(e))
         end
+        push!(factors, factor_term)
+        push!(powers, power_term)
     end
 
-    factor = isempty(factors) ? 1 : prod(factors)
-    powers, factor
-
+    factors, powers
 end
 
+function polynomial_propensity(expr, iv, smap, vars)
+    factor, powers = try
+        split_factor(expr, iv, smap, vars)
+    catch e
+        error("Propensity function $expr is non-polynomial? \n" * string(e))
+    end
 
-function polynomial_propensities(a::AbstractArray, iv::Sym, smap::Dict)
-
-    vars = [x for (x,_) in Base.sort(collect(smap), by=x->x[2])]
-    N = length(vars)
+    [ factor ], [ powers ]
+end
 
-    all_factors = copy.(fill([], size(a)))
-    all_powers = copy.(fill(Vector{Int}[], size(a)))
+function polynomial_propensities(arr::AbstractArray, iv::Sym, smap::AbstractDict)
+    # Do we need sort here?

```

Comment:
need to sort to make the variable order consistent with `speciesmap()` (assumed to be the case elsewhere)